### PR TITLE
Recipient balance check bugfix

### PR DESCRIPTION
**Pull request by nventuro**
This PR fixes a bug where the recipient's balance is checked after `preRelayedCall` is called, meaning balance draining during that call is not detected. I also added new tests (in a separate test file - future PRs will also use this new test suite) to check for all scenarios.

### CODE REVIEW

**Comment by drortirosh** [comment]
I think the balance check can be simplified, with the pre/post pattern:
1. calculate gasLimit. make sure balance is above. subtract gasLimit.
2. make the pre/relayed/post calls
3. calculate actual gas usage. refund balance with unused gas.

**Comment by nventuro** [comment]
That _would_ work, but we'd need at least one extra write to storage, so at a minimum 5k extra gas. Reading the balance only costs 400 gas.

**Comment by drortirosh** [comment]
> That _would_ work, but we'd need at least one extra write to storage, so at a minimum 5k extra gas. Reading the balance only costs 400 gas.

No need to use storage: we're in the context of our method (`recipientCallsAtomic`). its an in-memory variable

**Comment by nventuro** [comment]
Maybe I'm misunderstanding what you're proposing, could you share a short snippet of code describing your idea? I don't see how you could prevent the recipient from overwithdrawing without storing the subtracted value in store.

File: test/RelayHub.test.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
+const { BN, ether, expectEvent, time } = require('openzeppelin-test-helpers');
+
+const { getTransactionHash, getTransactionSignature } = require('../src/js/relayclient/utils')
+
+const RelayHub = artifacts.require('RelayHub');
+const SampleRecipient = artifacts.require('SampleRecipient');
+
+contract('RelayHub', function ([_, relayOwner, relay, sender, other]) {  // eslint-disable-line no-unused-vars
```
Comment (nventuro) [inline test/RelayHub.test.js]: We should be able to use the `varsIgnorePattern` option to avoid doing this, but I couldn't get it to work.

**Comment by drortirosh** [comment]
> Maybe I'm misunderstanding what you're proposing, could you share a short snippet of code describing your idea? I don't see how you could prevent the recipient from overwithdrawing without storing the subtracted value in store.

few lines before calling `recipientCallsAtomic`, we do 
```
require(gasPrice * initialGas <= balances[recipient], "Recipient balance too low");
```
so it should be followed by 
```
balances[recipient] -= gasPrice*initialGas
```
later, instead of `balances[recipient] -= charge`, we restore this "pre-charge", and charge the accurate amount:
`balances[recipient] += gasPrice*initialGas - charge`
(Note that the code assumes anyway `charge <  gasPrice*initialGas`, but now I read it again, and I think it might not be true, since there's the transactionfee in the `charge` calculation - but that's a separate issue to check)

**Comment by nventuro** [comment]
> few lines before calling recipientCallsAtomic, we do
>`require(gasPrice * initialGas <= balances[recipient], "Recipient balance too low");`

Wouldn't this require an additional `sstore`? That's what I meant in my initial reply. As far as I know, since [EIP 1087](https://eips.ethereum.org/EIPS/eip-1087) is not yet included, each `sstore` costs (at least) 5k gas.

>(Note that the code assumes anyway charge < gasPrice*initialGas, but now I read it again, and I think it might not be true, since there's the transactionfee in the charge calculation - but that's a separate issue to check)

Yes, we've already identified this bug and are currently finishing up a PR that tackles this.

**Comment by drortirosh** [comment]
> Wouldn't this require an additional `sstore`? 

why would it? we don't set any field, only stack-local variable.

**Comment by frangio** [comment]
`balances` is a mapping in storage, and `balances[recipient] -= gasPrice*initialGas` modifies an element in this mapping, so it results in an `SSTORE`.

**Comment by spalladino** [comment]
Thanks @drortirosh!

**Pull request merged**